File: leetcodecpp/leetcode/NineOneTwo.cpp
//
// Created by John on 2020/3/20.
//
#include <iostream>
#include "vector"
using namespace std;
class Solution{
public:
//    vector<int> sortArray(vector<int> & nums) {
//        if (nums.empty() || nums.size() == 0) {
//            return nums;
//        }
//
//    }
//    void quickSort(vector<int> & nums, int start, int end) {
//        if (start >= end) {
//            return;
//        }
//        int left = start, right = end;
//        int pivot = nums[start + end] / 2;
//        while (left <= right) {
//            while (left <= right && nums[left] < pivot) {
//                left++;
//            }
//            while (left <= right && nums[right] > pivot) {
//                right--;
//            }
//            if (left <= right) {
//                int temp = nums[left];
//                nums[left] = nums[right];
//                nums[right] = temp;
//                left++;
//                right--;
//            }
//        }
//        quickSort(nums, start, right);
//        quickSort(nums, left, end);
//    }
//    vector<int> sortArray(vector<int>& nums) {
//        if (nums.empty() || nums.size() == 0) {
//            return nums;
//        }
//        quickSort(nums, 0, nums.size() - 1);
//        return nums;
//    }
//    void quickSort(vector<int> & nums, int start, int end) {
//        if (start >= end) {
//            return;
//        }
//        int left = start;
//        int right = end;
//        int pivot = nums[(left + right) / 2];
//        while (left <= right) {
//            while (left <= right && nums[left] < pivot) {
//                left++;
//            }
//            while (left <= right && nums[right] > pivot) {
//                right--;
//            }
//            if (left <= right) {
//                int temp = nums[right];
//                nums[right] = nums[left];
//                nums[left] = temp;
//                left++;
//                right--;
//            }
//        }
//        quickSort(nums, start, right);
//        quickSort(nums, left, end);
//    }

//      vector<int> sortArray(vector<int>& nums) {
//            for (int i = 0; i < nums.size() - 1; i++) {
//                  for (int j = 0; j < nums.size() - i - 1; j++) {
//                        if (nums[j] > nums[j + 1]) {
//                              int temp = nums[j + 1];
//                              nums[j + 1] = temp;
//                              nums[j] = nums[j + 1];
//                        }
//                  }
//            }
//            return nums;
//      }
// public:
//     vector<int> sortArray(vector<int>& nums) {
//         quickSort(nums, 0, nums.size() - 1);
//         return nums;
//     }
//
//     void quickSort(vector<int>& nums, int startIndex, int endIndex) {
//         if (startIndex > endIndex) {
//             return;
//         }
//         int pivot = partition(nums, startIndex, endIndex);
//         quickSort(nums, startIndex, pivot - 1);
//         quickSort(nums, pivot + 1, endIndex);
//     }
//
//     int partition(vector<int>& nums, int startIndex, int endIndex) {
//         int pivot = nums[startIndex + rand() % (endIndex - startIndex + 1)];
//         int mark = startIndex;
//
//         for (int i = startIndex + 1; i <= endIndex; i++) {
//             if (nums[i] < pivot) {
//                 mark++;
//                 int p = nums[mark];
//                 nums[mark] = nums[i];
//                 nums[i] = p;
//             }
//         }
//
//         nums[startIndex] = nums[mark];
//         nums[mark] = pivot;
//
//         return mark;
//     }

public:
    vector<int> sortArray(vector<int>& nums) {
        bucketSort(nums, nums.size());
        return nums;
    }

    // Insertion sort function to sort individual buckets
    void insertionSort(vector<int>& bucket) {
        for (int i = 1; i < bucket.size(); i++) {
            int key = bucket[i];
            int j = i - 1;
            while (j >= 0 && bucket[j] > key) {
                bucket[j + 1] = bucket[j];
                j--;
            }
            bucket[j + 1] = key;
        }
    }

    void bucketSort(vector<int>& nums, int n) {
        int min_val = *min_element(nums.begin(), nums.end());
        int max_val = *max_element(nums.begin(), nums.end());
        int bucket_count = max_val - min_val + 1;


        // create bucket
        vector<vector<int>> bucket(bucket_count);

        // put array elements in different buckets
        for (int i = 0; i < n; i++) {
            int bi = nums[i] - min_val;
            bucket[bi].push_back(nums[i]);
        }

        // Sort individual buckets by insertion sort
        for (int i = 0; i < bucket_count; i++) {
            insertionSort(bucket[i]);
        }

        // Concatenate all buckets into arr[]
        int index = 0;
        for (int i = 0; i < bucket_count; i++) {
            for (int j = 0; j < bucket[i].size(); j++) {
                nums[index++] = bucket[i][j];
            }
        }
    }


};

int main() {
    vector<int> nums = {-2, -3, 5};

    Solution * solution = new Solution();
    solution->sortArray(nums);
}

File: leetcodejava/src/main/java/com/leetcode/NineOneTwo.java
package com.leetcode;

/**
 * @author: John
 * @date: 2020-09-24 09:16
 * @description: 912
 **/
public class NineOneTwo {
    public void sort(int arr[]) {
//        for (int i = 0; i < arr.length - 1; i++) {
//            for (int j = 0; j < arr.length - 1 - i; j++) {
//                if (arr[j] > arr[j + 1]) {
//                    int temp = arr[j];
//                    arr[j] = arr[j+1];
//                    arr[j+1] = temp;
//                }
//            }
//        }
        quickSort(arr, 0, arr.length - 1);
    }



    public void quickSort(int[] arr, int l, int r) {
        // 递归结束条件: l 大于等于 r
        if (l >= r) {
           return;
        }
        int pivot = partition(arr, l, r);
        quickSort(arr, l, pivot - 1);
        quickSort(arr, pivot + 1, r);
    }

    private int partition(int[] arr, int l, int r) {
        int pivot = arr[l];
        int i = l;
        int j = r;

        while (i != j) {
            // 右指针左移
            while (i < j && arr[j] > pivot) {
                j--;
            }
            // 左指针右移
            while (i < j && arr[i] <= pivot) {
                i++;
            }
            // swap element of l and r
            if (i < j) {
                int p = arr[i];
                arr[i] = arr[j];
                arr[j] = p;
            }
        }
        arr[l] = arr[i];
        arr[i] = pivot;
        return i;
    }

    public static void main(String[] args) {
//        int[] n = {5,2,3,1};
//        NineOneTwo nineOneTwo = new NineOneTwo();
//        nineOneTwo.sort(n);
////        System.out.println();
//        for (int i = 0; i < n.length; i++) {
//            System.out.println(n[i]);
//        }

        System.out.println( 1 << 30);
    }
}
